File: authserver/src/main/java/l2s/authserver/network/gamecomm/as2gs/LoginServerFail.java
package l2s.authserver.network.gamecomm.as2gs;

import io.netty.buffer.ByteBuf;
import l2s.authserver.network.gamecomm.SendablePacket;

public class LoginServerFail extends SendablePacket {
    private final String _reason;
    private final boolean _restartConnection;

    public LoginServerFail(String reason, boolean restartConnection) {
        _reason = reason;
        _restartConnection = restartConnection;
    }

    @Override
    protected ByteBuf writeImpl(ByteBuf byteBuf) {
        byteBuf.writeByte(0);
        writeString(byteBuf, _reason);
        byteBuf.writeByte(_restartConnection ? 1 : 0);
        return byteBuf;
    }

    @Override
    protected byte getOpCode() {
        return 1;
    }
}

File: authserver/src/main/java/l2s/authserver/network/gamecomm/as2gs/LoginServerFail.kt
package l2s.authserver.network.gamecomm.as2gs

import io.netty.buffer.ByteBuf
import l2s.authserver.network.gamecomm.SendablePacket

/**
 * @reworked by Bonux
 */
class LoginServerFail(private val reason: String, private val restartConnection: Boolean) : SendablePacket() {

    override val opCode: Byte = 0x01

    override fun writeImpl(byteBuf: ByteBuf): ByteBuf {
        byteBuf.writeByte(0x00) // Reason ID

        writeString(byteBuf, reason)
        byteBuf.writeByte(if (restartConnection) 0x01 else 0x00)

        return byteBuf
    }

}